ndcube/ndcube.py

```diff
@@ -443,31 +445,47 @@ def quantity(self):
         """Unitful representation of the NDCube data."""
         return u.Quantity(self.data, self.unit, copy=_NUMPY_COPY_IF_NEEDED)
 
-    def _generate_world_coords(self, pixel_corners, wcs, needed_axes=None, *, units):
-        # Create meshgrid of all pixel coordinates.
-        # If user wants pixel_corners, set pixel values to pixel corners.
+
+    def _generate_independent_world_coords(self, pixel_corners, wcs, pixel_axes, units):
+        naxes = len(self.data.shape)
+        pixel_indices = [np.array([0], dtype=int).reshape([1] * naxes).squeeze()] * naxes
+        for pixel_axis in pixel_axes:
+            len_axis = self.data.shape[::-1][pixel_axis]
+            # Define limits of desired pixel range based on whether corners or centers are desired
+            lims = (-0.5, len_axis + 1) if pixel_corners else (0, len_axis)
+            pix_ind = np.arange(lims[0], lims[1])
+            shape = [1] * naxes
+            shape[pixel_axis] = len(pix_ind)
+            pixel_indices[pixel_axis] = pix_ind.reshape(shape)
+        world_coords = wcs.pixel_to_world_values(*pixel_indices)
+        # TODO: Remove NaNs??? These should not be here
+        if np.isnan(world_coords).any():
+            if isinstance(world_coords, tuple| list):
+                world_coords = [world_coord[~np.isnan(world_coord)] for world_coord in world_coords]
+            else:
+                world_coords = world_coords[~np.isnan(world_coords)]
+        if units:
+            mod = abs(wcs.world_n_dim - naxes) if wcs.world_n_dim > naxes else 0
```

Comment:
This was my attempt at working out if there is a difference between world dims and pixel dims.